includes/validation/class-amp-validation-error-taxonomy.php

```diff
@@ -3129,6 +3129,101 @@ static function ( $attribute_name ) {
+				return sprintf(
+					__( "Tag '%1\$s' is disallowed as first child of tag '%2\$s'", 'amp' ),
+					esc_html( $validation_error['first_child_tag'] ),
+					esc_html( $validation_error['node_name'] )
+				);
+
+			case AMP_Tag_And_Attribute_Sanitizer::INCORRECT_NUM_CHILD_TAGS:
+				/* translators: %1$s is the node name, %2$s is required child count */
+				return sprintf(
+					_n( "Tag '%1\$s' must have %2\$d child tag", "Tag '%1\$s' must have %2\$d child tags", (int) $validation_error['required_child_count'], 'amp' ),
+					esc_html( $validation_error['node_name'] ),
+					esc_html( $validation_error['required_child_count'] )
+				);
+
+			case AMP_Tag_And_Attribute_Sanitizer::INCORRECT_MIN_NUM_CHILD_TAGS:
+				/* translators: %1$s is the node name, %2$s is required child count */
+				return sprintf(
+					_n( "Tag '%1\$s' must have a minimum of %2\$d child tag", "Tag '%1\$s' must have a minimum of %2\$d child tags", (int) $validation_error['required_min_child_count'], 'amp' ),
+					esc_html( $validation_error['node_name'] ),
+					esc_html( $validation_error['required_min_child_count'] )
+				);
+
+			case AMP_Tag_And_Attribute_Sanitizer::WRONG_PARENT_TAG:
+				/* translators: %1$s is the node name, %2$s is parent name */
+				return sprintf(
+					__( "The parent tag of tag '%1\$s' cannot be '%2\$s'", 'amp' ),
+					esc_html( $validation_error['node_name'] ),
+					esc_html( $validation_error['parent_name'] )
+				);
+
+			case AMP_Tag_And_Attribute_Sanitizer::DISALLOWED_TAG_ANCESTOR:
+				/* translators: %1$s is the node name, %2$s is the disallowed ancestor tag name */
+				return sprintf(
+					__( "The tag '%1\$s' may not appear as a descendant of tag '%2\$s'", 'amp' ),
+					esc_html( $validation_error['node_name'] ),
+					esc_html( $validation_error['disallowed_ancestor'] )
+				);
+
+			case AMP_Tag_And_Attribute_Sanitizer::MANDATORY_TAG_ANCESTOR:
+				/* translators: %1$s is the node name, %2$s is the required ancestor tag name */
+				return sprintf(
+					__( "The tag '%1\$s' may only appear as a descendant of tag '%2\$s'", 'amp' ),
+					esc_html( $validation_error['node_name'] ),
+					esc_html( $validation_error['required_ancestor_name'] )
+				);
+
+			case AMP_Tag_And_Attribute_Sanitizer::INVALID_ATTR_VALUE:
+				/* translators: %1$s is the attribute name, %2$s is the element name, %3$s is the invalid attribute value */
+				return sprintf(
+					__( "The attribute '%1\$s' in tag '%2\$s' is set to the invalid value '%3\$s'", 'amp' ),
+					esc_html( $validation_error['node_name'] ),
+					esc_html( $validation_error['parent_name'] ),
+					esc_html( $validation_error['element_attributes'][ $validation_error['node_name'] ] )
+				);
+
+			case AMP_Tag_And_Attribute_Sanitizer::INVALID_URL_PROTOCOL:
+				$parsed_url = wp_parse_url( $validation_error['element_attributes'][ $validation_error['node_name'] ] );
+				$invalid_protocol = isset( $parsed_url['scheme'] ) ? $parsed_url['scheme'] : '';
```

Comment:
Not sure what to do here if the scheme could not be parsed, defaulting to an empty string for now.

Reply:
Maybe `null` would be more appropriate?

Reply:
Oh I see, it's used in a string. Yeah, an empty string seems fine for now.

Reply:
Actually, since `code` is now being used as opposed to quotes, then I think something else is needed:

```suggestion
				$invalid_protocol = isset( $parsed_url['scheme'] ) ? $parsed_url['scheme'] . ':' : '(failure)';
```